src/control/lib/spdk/src/nvme_control_common.c

```diff
@@ -234,19 +250,49 @@ _discover(prober probe, bool detach, health_getter get_health)
 }
 
 static int
-copy_ctrlr_data(struct ctrlr_t *cdst, const struct spdk_nvme_ctrlr_data *cdata)
+str2ctrlr(char **dst, const void *src)
 {
-	if (copy_ascii(cdst->model, sizeof(cdst->model), cdata->mn,
-		       sizeof(cdata->mn)) != 0)
-		return -NVMEC_ERR_CHK_SIZE;
+	int len;
+
+	assert(src != NULL);
+	assert(dst != NULL);
+	assert(*dst == NULL);
 
-	if (copy_ascii(cdst->serial, sizeof(cdst->serial), cdata->sn,
-		       sizeof(cdata->sn)) != 0)
+	len = strnlen((const char *)src, NVME_DETAIL_BUFLEN);
+	if (len == NVME_DETAIL_BUFLEN) {
+		perror("src buf too big");
 		return -NVMEC_ERR_CHK_SIZE;
+	}
+
+	*dst = calloc(1, len + 1);
+	if (*dst == NULL)
+		return -ENOMEM;
 
-	if (copy_ascii(cdst->fw_rev, sizeof(cdst->fw_rev), cdata->fr,
-		       sizeof(cdata->fr)) != 0)
+	if (copy_ascii(*dst, len + 1, src, len) != 0) {
+		perror("copy_ascii");
 		return -NVMEC_ERR_CHK_SIZE;
```

Comment:
memory is freed on error through `free_ctrlr_fields()`

Reply:
Since it's a static function with a pretty limited scope, that seems okay to me. It makes it a bit harder to keep track mentally of the state of the memory, though.